File: lec0_intro1/StringMemory.java
package lec1_intro1;

public class StringMemory {
    public static void main(String[] args) {
        // In Java, memory is automatically managed!
        String greeting = "Hello, CSE 105!";
        System.out.println(greeting);
        
        // No need to free memory - Java's garbage collector does it for you!
    }
}

/*
MEMORY MANAGEMENT IN JAVA:
1. Memory is automatically allocated when you create objects.
2. No need to check if allocation succeeded - Java handles it.
3. No need to manually free memory - garbage collector does it!
4. The garbage collector finds and frees unused objects.
5. You can't accidentally use freed memory (no dangling pointers).
6. This makes Java safer and easier, but slightly slower than C.
*/

File: lec0_intro1/StringMemory.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    // In C, we must manually allocate memory for strings
    char *greeting = malloc(20 * sizeof(char));
    
    if (greeting == NULL) {
        printf("Memory allocation failed!\n");
        return 1;
    }
    
    strcpy(greeting, "Hello, CSE 105!");
    printf("%s\n", greeting);
    
    // MUST remember to free memory or we have a memory leak!
    free(greeting);
    
    return 0;
}

/*
MEMORY MANAGEMENT IN C:
1. You must manually allocate memory using malloc().
2. You must check if allocation succeeded (NULL check).
3. You must manually free memory using free().
4. Forgetting to free causes memory leaks.
5. Using freed memory causes crashes (dangling pointers).
6. This is YOUR responsibility - the compiler won't help!
*/
